akeneo-design-system/src/components/Pagination/Pagination.tsx

```diff
@@ -0,0 +1,181 @@
+`;
+
+type PaginationItemProps = {
+  currentPage: boolean;
+  onClick: (page: number) => void;
+};
+
+const PaginationItem: FC<PaginationItemProps> = ({currentPage, children, onClick}) => {
+  const title = `No. ${children as string}`;
+
+  return (
+    <PaginationItemContainer
+      onClick={() => (children == PAGINATION_SEPARATOR ? null : onClick(children as number))}
+      data-testid="paginationItem"
+      title={children != PAGINATION_SEPARATOR ? title : ''}
+      disabled={children == PAGINATION_SEPARATOR}
+      currentPage={currentPage}
+    >
+      {children}
+    </PaginationItemContainer>
+  );
+};
+
+const currentPaginationItemMixin = css`
+  border: 1px ${getColor('brand', 100)} solid;
+  color: ${getColor('brand', 100)};
+`;
+
+const otherPaginationItemMixin = css`
+  border: 1px ${getColor('grey', 80)} solid;
+  color: ${getColor('grey', 100)};
+`;
+
+const disabledMixin = css`
+  cursor: default;
+  :hover {
+    background-color: ${getColor('white')};
+  }
+`;
+
+const PaginationItemContainer = styled.span<AkeneoThemedProps & PaginationItemProps & {disabled: boolean}>`
+  ${props => (props.currentPage ? currentPaginationItemMixin : otherPaginationItemMixin)}
+  display: inline-block;
+  border-width: 1px;
+  font-size: 13px;
+  font-weight: 400;
+  text-transform: uppercase;
+  border-radius: 16px;
+  padding: 0 10px;
+  height: 22px;
+  line-height: 21px;
+  cursor: pointer;
+  font-family: inherit;
+  transition: background-color 0.1s ease 0s;
+  min-width: 40px;
+  text-align: center;
+  box-sizing: border-box;
+
+  :not(:last-child) {
+    margin-right: 10px;
+  }
+
+  :hover {
+    background-color: ${getColor('grey', 20)};
+  }
+
+  ${props => (props.disabled ? disabledMixin : null)}
+`;
+
+function computePages(currentPage: number, numberOfPages: number) {
+  if (numberOfPages <= MAX_PAGINATION_ITEMS_WITHOUT_SEPARATOR) {
+    return Array.from(Array(numberOfPages).keys()).map((page: number) => page + 1);
+  }
+
+  const FIRST_PAGE = 1;
+  const SECOND_PAGE = 2;
+  const THIRD_PAGE = 3;
+  const FOURTH_PAGE = 4;
+  const LAST_PAGE = numberOfPages;
+  const SECOND_LAST = LAST_PAGE - 1;
+  const THIRD_LAST = LAST_PAGE - 2;
+  const FOURTH_LAST = LAST_PAGE - 3;
+  const PREVIOUS_PAGE = currentPage - 1;
+  const NEXT_PAGE = currentPage + 1;
+
+  const pages: Array<number | string> = [FIRST_PAGE];
+
+  if (currentPage >= FOURTH_PAGE) {
+    pages.push(PAGINATION_SEPARATOR);
+  }
+
+  if (currentPage > SECOND_PAGE) {
+    if (currentPage === LAST_PAGE) {
+      pages.push(THIRD_LAST);
+    }
+    pages.push(PREVIOUS_PAGE);
+  }
+
+  if (currentPage !== FIRST_PAGE && currentPage !== LAST_PAGE) {
+    pages.push(currentPage);
+  }
+
+  if (currentPage < SECOND_LAST) {
+    pages.push(NEXT_PAGE);
+  }
+
+  if (currentPage === FIRST_PAGE) {
+    pages.push(THIRD_PAGE);
+  }
+
+  if (currentPage <= FOURTH_LAST) {
+    pages.push(PAGINATION_SEPARATOR);
+  }
+
+  pages.push(LAST_PAGE);
+
+  return pages;
```

Comment:
What do you think about:
```suggestion
  const before = [
    1,
    2,
    3,
    4,
    currentPage - 1
  ];
  
  const after = [
    currentPage + 1
    numberOfPages - 3,
    numberOfPages - 2,
    numberOfPages - 1,
    numberOfPages
  ]
  
  const separator = currentPage >= 4 || currentPage <= numberOfPages - 3 ? 
    [PAGINATION_SEPARATOR] : 
    [currentPage];
  
  return arrayUnique([...before, ...separator, ...after])
}

const arrayUnique = (array: unknown[]) => {
  return [...new Set(array)]
```

Reply:
I've just tried yours and it doesn't pass the tests.

So I got curious and I tried to make a short one with reduce, but it's not as readable as I expected. Anyway,
```
return Array.from(new Array(total), (x, i) => i + 1)
  .reduce((pages, page) => {
    if (
      page === 1 ||
      page === total ||
      page === current ||
      page === current - 1 ||
      page === current + 1 ||
      (current === 1 && page === 3) ||
      (current === total && page === total - 2)
    ) {
      pages.push(page);
    }
  
    if (
      (page === current - 2 && page > 1 && current < total) ||
      (page === current - 3 && current === total) ||
      (page === current + 2 && page < total && total > 4)
    ) {
      pages.push('...');
    }
  
    return pages;
  }, []);
```

Reply:
By viewing the screenshot, I realize that my implem is not right. But my point was more about the fact to avoid multiple ifs and use array instead. Maybe, I'm wrong and it's more complicated

Reply:
we're not proud of our `if` implementation, but it's the more "human readable" we've found.